File: rust/ximea_camera/src/messages.rs
use log::{info, warn};
use zmq::{Context, Socket, SocketType};

pub struct Publisher {
    pub_socket: Socket,
    rep_socket: Socket,
}

impl Publisher {
    pub fn new(ctx: &Context, pub_port: u16, handshake_port: u16) -> Self {
        let pub_socket = ctx.socket(SocketType::PUB).unwrap();
        pub_socket.bind(&format!("tcp://*:{pub_port}")).unwrap();

        let rep_socket = ctx.socket(SocketType::REP).unwrap();
        rep_socket
            .bind(&format!("tcp://*:{handshake_port}"))
            .unwrap();

        Publisher {
            pub_socket,
            rep_socket,
        }
    }

    pub fn wait_for_subscriber(&self) {
        loop {
            match self.rep_socket.recv_string(0) {
                Ok(Ok(msg)) if &msg == "Hello" => {
                    self.rep_socket.send("Welcome", 0).unwrap();
                    break;
                }
                Ok(Err(e)) => warn!("Error receiving message: {:?}", e),
                Err(e) => warn!("Communication error: {}", e),
                _ => {}
            }
        }
    }

    pub fn publish(&self, topic: &str, msg: &str) {
        let full_msg = format!("{} {}", topic, msg);
        self.pub_socket.send(full_msg.as_bytes(), 0).unwrap();
        info!("Published message on topic '{}': {}", topic, msg);
    }
}

pub struct Subscriber {
    sub_socket: Socket,
    req_socket: Socket,
}

impl Subscriber {
    pub fn new(ctx: &Context, pub_port: u16, handshake_port: u16, server_ip: &str) -> Self {
        let sub_socket = ctx.socket(SocketType::SUB).unwrap();
        sub_socket
            .connect(&format!("tcp://{server_ip}:{pub_port}"))
            .unwrap();

        let req_socket = ctx.socket(SocketType::REQ).unwrap();
        req_socket
            .connect(&format!("tcp://{server_ip}:{handshake_port}"))
            .unwrap();

        Subscriber {
            sub_socket,
            req_socket,
        }
    }

    pub fn handshake(&self) {
        self.req_socket.send("Hello", 0).unwrap();
        match self.req_socket.recv_string(0) {
            Ok(Ok(reply)) if &reply == "Welcome" => info!("Handshake successful."),
            Ok(Ok(_)) => warn!("Unexpected reply during handshake."),
            Ok(Err(e)) => warn!("Error receiving reply: {:?}", e),
            Err(e) => warn!("Communication error during handshake: {}", e),
        }
    }

    pub fn subscribe(&self, topic: &str) {
        self.sub_socket.set_subscribe(topic.as_bytes()).unwrap();
        info!("Subscribed to topic '{}'.", topic);
    }

    pub fn receive(&self) -> String {
        let msg = self.sub_socket.recv_string(0).unwrap().unwrap();
        info!("Received message: {}", msg);
        msg
    }
}

File: modules/messages.py
import time

import zmq
from modules.utils.log_config import setup_logging

logger = setup_logging(logger_name="Messages", level="INFO", color="cyan")


class Publisher:
    def __init__(self, pub_port, handshake_port):
        self.context = zmq.Context()

        # Publisher socket
        self.pub_socket = self.context.socket(zmq.PUB)
        self.pub_socket.bind(f"tcp://*:{pub_port}")

        # Handshake socket
        self.rep_socket = self.context.socket(zmq.REP)
        self.rep_socket.bind(f"tcp://*:{handshake_port}")

    def wait_for_subscriber(self):
        # Wait for a handshake message from subscriber
        try:
            message = self.rep_socket.recv_string()  # Non-blocking receive
            if message == "Hello":
                self.rep_socket.send_string("Welcome")
                logger.info("Handshake completed with a subscriber.")
        except zmq.Again as e:
            logger.warning(f"{e} No handshake request received yet.")

    def publish(
        self,
        msg,
        topic="",
    ):
        self.pub_socket.send_string(f"{topic} {msg}")
        logger.debug(f"Published message on topic '{topic}': {msg}")

    def close(self):
        self.pub_socket.close()
        self.rep_socket.close()
        self.context.term()
        logger.info("Closed publisher sockets and terminated context.")


class Subscriber:
    def __init__(
        self,
        pub_port,
        handshake_port,
        server_ip="localhost",
        retry_attempts=5,
        retry_timeout=5,
    ):
        self.context = zmq.Context()
        self.retry_attempts = retry_attempts
        self.retry_timeout = retry_timeout

        # Subscriber socket
        self.sub_socket = self.context.socket(zmq.SUB)
        self.sub_socket.connect(f"tcp://{server_ip}:{pub_port}")

        # Handshake socket
        self.req_socket = self.context.socket(zmq.REQ)
        self.req_socket.connect(f"tcp://{server_ip}:{handshake_port}")

    def handshake(self):
        for _ in range(self.retry_attempts):
            try:
                # Send handshake message
                self.req_socket.send_string("Hello")
                reply = self.req_socket.recv_string()
                if reply == "Welcome":
                    logger.info("Handshake successful.")
                    return True
                else:
                    logger.warning("Handshake failed. Unexpected reply.")
            except zmq.ZMQError as e:
                logger.error(f"Handshake error: {e}")
            time.sleep(self.retry_timeout)  # wait before retrying
        logger.error("Handshake failed after maximum retry attempts.")

        return False

    def subscribe(self, topic=""):
        self.sub_socket.setsockopt_string(zmq.SUBSCRIBE, topic)
        logger.info(f"Subscribed to topic '{topic}'.")

    def receive(self, block=False):
        try:
            # Receive the message, with or without blocking based on the 'block' parameter
            message = self.sub_socket.recv_string(flags=0 if block else zmq.NOBLOCK)

            # Split the message into topic and actual message
            topic, actual_message = message.split(" ", 1)
            logger.debug(f"Received message: {message}")

            return topic, actual_message

        except zmq.Again:
            # logger.debug("No message received yet: %s", e)
            return None, None

        except ValueError as e:
            logger.error("Error parsing message: %s", e)
            return None, None

        except Exception as e:
            logger.error("Unexpected error: %s", e)
            return None, None

    def close(self):
        self.sub_socket.close()
        self.req_socket.close()
        self.context.term()
        logger.info("Closed subscriber sockets and terminated context.")
